File: Object/Shape/TriangleMesh.cpp
#include "Object/Shape/TriangleMesh.hpp"
#include "Object/Shape/Triangle.hpp"

#include <algorithm>

namespace Object {
    namespace Shape {
        TriangleMesh::TriangleMesh(std::vector<Vertex> &&vertices, std::vector<Triangle> &&triangles)
            : mVertices(std::move(vertices)), mTriangles(std::move(triangles)), mBoundingVolumeHierarchy(computeBoundingVolumeHierarchy())
        {
        }

        TriangleMesh::TriangleMesh(std::vector<Vertex> &&vertices, std::vector<Triangle> &&triangles, Object::BoundingVolumeHierarchy &&boundingVolumeHierarchy)
            : mVertices(std::move(vertices)), mTriangles(std::move(triangles)), mBoundingVolumeHierarchy(std::move(boundingVolumeHierarchy))
        {
        }

        bool TriangleMesh::intersect(const Math::Ray &ray, Intersection &isect, bool closest) const
        {
            BoundingVolume::RayData rayData = BoundingVolume::getRayData(ray);

            auto callback = [&](unsigned int index, float &) {
                const Triangle &triangle = mTriangles[index];
                bool ret = false;

                const Vertex &vertex0 = mVertices[triangle.vertices[0]];
                const Vertex &vertex1 = mVertices[triangle.vertices[1]];
                const Vertex &vertex2 = mVertices[triangle.vertices[2]];

                float tu, tv;
                if (Object::Shape::Triangle::intersect(ray, vertex0.point, vertex1.point, vertex2.point, isect.distance, tu, tv)) {
                    isect.normal = triangle.normal;
                    isect.tangent = Math::Bivector(Math::Vector(), Math::Vector());
                    isect.surfacePoint = Math::Point2D();
                    ret = true;
                }
                return ret;
            };

            return mBoundingVolumeHierarchy.intersect(rayData, isect.distance, closest, std::ref(callback));
        }

        BoundingVolume TriangleMesh::boundingVolume(const Math::Transformation &trans) const
        {
            BoundingVolume volume;
            for (const Vertex &vertex : mVertices) {
                volume.expand(trans * vertex.point);
            }

            return volume;
        }

        const Object::BoundingVolumeHierarchy &TriangleMesh::boundingVolumeHierarchy() const
        {
            return mBoundingVolumeHierarchy;
        }

        void TriangleMesh::writeProxy(ShapeProxy &proxy, OpenCL::Allocator &clAllocator) const
        {
            proxy.type = ShapeProxy::Type::TriangleMesh;
            proxy.triangleMesh.vertices = clAllocator.allocateArray<PointProxy>(mVertices.size());
            for(int i=0; i<mVertices.size(); i++) {
                mVertices[i].point.writeProxy(proxy.triangleMesh.vertices[i]);
            }
            proxy.triangleMesh.triangles = clAllocator.allocateArray<TriangleProxy>(mTriangles.size());
            for(int i=0; i<mTriangles.size(); i++) {
                for(int j=0; j<3; j++) {
                    proxy.triangleMesh.triangles[i].vertices.values[j] = mTriangles[i].vertices[j];
                    mTriangles[i].normal.writeProxy(proxy.triangleMesh.triangles[i].normal);
                }
            }
            proxy.triangleMesh.bvh = clAllocator.allocateArray<BVHNodeProxy>(mBoundingVolumeHierarchy.nodes().size());
            mBoundingVolumeHierarchy.writeProxy(proxy.triangleMesh.bvh);
        }

        Object::BoundingVolumeHierarchy TriangleMesh::computeBoundingVolumeHierarchy() const
        {
            std::vector<Math::Point> centroids(mTriangles.size());
            for (unsigned int i = 0; i < mTriangles.size(); i++) {
                const Triangle &triangle = mTriangles[i];
                Math::Vector centroid;
                for (unsigned int j = 0; j < 3; j++) {
                    centroid = centroid + Math::Vector(mVertices[triangle.vertices[j]].point);
                }
                centroids[i] = Math::Point(centroid / 3.0f);
            }

            auto func = [&](int index) {
                BoundingVolume volume;
                for (unsigned int i = 0; i < 3; i++) {
                    volume.expand(mVertices[mTriangles[index].vertices[i]].point);
                }
                return volume;
            };

            return Object::BoundingVolumeHierarchy(centroids, std::ref(func));
        }
    }
}

File: src/object/shape/triangle_mesh.rs
use crate::geo;
use crate::object;

use geo::Bivec3;
use geo::Normal3;
use geo::Point2;
use geo::Point3;
use geo::Transformation;

use object::BoundingVolume;
use object::BoundingVolumeHierarchy;
use object::Shape;
use object::ShapeIntersection;

pub struct Vertex {
    pub point: Point3,
}

pub struct Triangle {
    pub vertices: [usize; 3],
    pub normal: Normal3,
}

pub struct TriangleMesh {
    vertices: Vec<Vertex>,
    triangles: Vec<Triangle>,
    bvh: BoundingVolumeHierarchy,
}

impl TriangleMesh {
    fn compute_bvh(
        vertices: &[Vertex],
        triangles: &[Triangle],
    ) -> BoundingVolumeHierarchy {
        let mut centroids = Vec::with_capacity(triangles.len());
        for triangle in triangles {
            let mut centroid = Point3::ZERO;
            for j in 0..3 {
                centroid += vertices[triangle.vertices[j]].point;
            }
            centroids.push(centroid / 3.0);
        }

        let func = |index: usize| {
            let mut volume = BoundingVolume::new();
            for i in 0..3 {
                volume.include_point(vertices[triangles[index].vertices[i]].point);
            }
            return volume;
        };

        return BoundingVolumeHierarchy::from_volumes(&centroids[..], &func);
    }

    pub fn new(
        vertices: Vec<Vertex>,
        triangles: Vec<Triangle>,
    ) -> TriangleMesh {
        let bvh = Self::compute_bvh(&vertices[..], &triangles[..]);
        return Self::new_with_bvh(vertices, triangles, bvh);
    }

    pub fn new_with_bvh(
        vertices: Vec<Vertex>,
        triangles: Vec<Triangle>,
        bvh: BoundingVolumeHierarchy,
    ) -> TriangleMesh {
        return TriangleMesh {
            vertices,
            triangles,
            bvh,
        };
    }
}

impl Shape for TriangleMesh {
    fn bounding_volume(&self, xform: &Transformation) -> BoundingVolume {
        let mut volume = BoundingVolume::new();

        for vertex in &self.vertices {
            volume.include_point(vertex.point.transform(xform));
        }

        return volume;
    }

    fn intersect(
        &self,
        ray: geo::Ray,
        max_distance: f32,
        closest: bool,
    ) -> Option<ShapeIntersection> {
        let raydata = BoundingVolume::get_raydata(ray);
        let mut shape_isect = None;

        let mut func = |index: usize, max_distance: f32| {
            let triangle = &self.triangles[index];

            let vertex0 = &self.vertices[triangle.vertices[0]];
            let vertex1 = &self.vertices[triangle.vertices[1]];
            let vertex2 = &self.vertices[triangle.vertices[2]];

            if let Some((_tu, _tv, d)) = object::shape::Triangle::intersect(
                ray,
                vertex0.point,
                vertex1.point,
                vertex2.point,
                max_distance,
            ) {
                shape_isect = Some(ShapeIntersection::new(
                    d,
                    triangle.normal,
                    Bivec3::ZERO,
                    Point2::ZERO,
                ));
                return Some(d);
            } else {
                return None;
            }
        };

        self.bvh
            .intersect(raydata, max_distance, closest, &mut func);
        return shape_isect;
    }
}
